### ISSUE TITLE
Forcing celery workers to not sleep or improving ETA

### ISSUE THREAD

**Issue by legraina** [celery/celery#4191]
Hi,

## Checklist

I'm using the stable celery 4.0.2 (installed with pip) and here my report:

software -> celery:4.0.2 (latentcall) kombu:4.0.2 py:3.4.2
            billiard:3.5.0.2 py-amqp:2.1.4
platform -> system:Linux arch:64bit, ELF imp:CPython
loader   -> celery.loaders.app.AppLoader
settings -> transport:amqp results:disabled

task_queues: {
    'queue1': {   'binding_key': '********',
                               'exchange': 'queue1'},
    'queue2': {'binding_key': '********', 'exchange': 'queue2'}}
cfg: <module 'celery_workers.workersconfig' from '/usr/src/workers/celery_workers/workersconfig.py'>
key: '********'
beat_max_loop_interval: 60
enable_utc: True
os: <module 'os' from '/usr/lib/python3.4/os.py'>
value: 'sequential_backend_north'
accept_content: ['json']
task_ignore_result: True
task_serializer: 'json'
result_serializer: 'json'
task_routes: {
    'celery_workers.workers.queue1.*': {   'queue': 'queue1'},
    'celery_workers.workers.queue2.*': {'queue': 'queue2'}}
broker_url: 'amqp://guest:********@rabbitmq:5672//'
timezone: 'US/Eastern'
worker_timer_precision: 0.01
crontab: <class 'celery.schedules.crontab'>
include: ['celery_workers.workers.queue1',
 'celery_workers.workers.queue2']

## Steps to reproduce

I'm trying to simulate the system before deployment. To do so, I'm simulating the time with a task simulate_time that updates the time second by second: it updates the db with the time and apply_async an another simulate_time with time+1. Then, I play with the speed of the system: I set the ETA of the next simulate_time with datetime.utcnow()+timedelta(microseconds=round(1e6/speed)).

## Expected behavior

I was wondering if there is a way of forcing workers to not sleep or waking up faster such as to run the task on the expected ETA (at least in a development mode setting).

## Actual behavior

Celery cannot follow a speed of 16. It seems that the workers go to sleep and take some time before waking up when receiving a new task, thus slowing the overall simulation. I've attached below a log for a speed of 16. If the simulation was working correctly, we should see around 16 events within a minute instead of only 4. And if we continue of increasing the speed, celery cannot follow.
For example, we can see (in the log below) that the task is sent at 14:01:38,308, received at 14:01:38,314 with an ETA of 10:01:38.366330, but only processed at 14:01:38,495 with a processing time of 0.012 (a waste of 0.12 s.).

A log:
disp-simulator-workers | [2017-08-10 14:01:38,056: INFO/PoolWorker-2] Task celery_workers.workers.simulator.simulate_time[a17f220d-2ee6-46db-9e46-87a91b5ebe0c] succeeded in 0.006356447993312031s: 40702
disp-simulator-workers | [2017-08-10 14:01:38,070: INFO/MainProcess] Received task: celery_workers.workers.simulator.simulate_time[3411b0f1-cce8-4634-ad40-122177e3ecc1]  ETA:[2017-08-10 10:01:38.114537-04:00] 
disp-simulator-workers | [2017-08-10 14:01:38,308: INFO/PoolWorker-2] Task celery_workers.workers.simulator.simulate_time[3411b0f1-cce8-4634-ad40-122177e3ecc1] succeeded in 0.006479467963799834s: 40703
disp-simulator-workers | [2017-08-10 14:01:38,312: INFO/MainProcess] Received task: celery_workers.workers.simulator.simulate_time[81924d54-2e96-4636-807f-eee48f444062]  ETA:[2017-08-10 10:01:38.366330-04:00] 
disp-simulator-workers | [2017-08-10 14:01:38,495: INFO/PoolWorker-1] Task celery_workers.workers.simulator.simulate_time[b1691278-90c0-4dcf-8f23-b64cfd40f936] succeeded in 0.011639489035587758s: 40704
disp-simulator-workers | [2017-08-10 14:01:38,500: INFO/MainProcess] Received task: celery_workers.workers.simulator.simulate_time[c15fdd13-5824-428d-9dc1-b67cce61f6f4]  ETA:[2017-08-10 10:01:38.551918-04:00] 
disp-simulator-workers | [2017-08-10 14:01:38,561: INFO/PoolWorker-1] Task celery_workers.workers.simulator.simulate_time[c15fdd13-5824-428d-9dc1-b67cce61f6f4] succeeded in 0.005412194004748017s: 40705
disp-simulator-workers | [2017-08-10 14:01:38,566: INFO/MainProcess] Received task: celery_workers.workers.simulator.simulate_time[bd693495-6117-4010-8508-f9e11e7a23ba]  ETA:[2017-08-10 10:01:38.620754-04:00]

**Comment by auvipy** [comment]
sorry for late response could you verify the issue on top of latest master?

**Comment by legraina** [comment]
Same issue on master.

Here, my report:
software -> celery:4.1.0 (latentcall) kombu:4.1.0 py:3.5.2 billiard:3.5.0.3 redis:2.10.5
platform -> system:Linux arch:64bit, ELF imp:CPython
loader   -> celery.loaders.app.AppLoader
settings -> transport:redis results:disabled
worker_timer_precision: 0.1
accept_content: ['json']


Here the function I'm testing:
```python
def simulate_time(time):
    eta = datetime.utcnow() + timedelta(microseconds=1e6/16)
    simulate_time.apply_async([time+1], eta=eta)
```

Here, the trace:
[2018-01-02 16:38:16,234: INFO/MainProcess] Received task: celery_workers.workers.simulator.simulate_time[ae12d705-9c7d-4cb5-9192-2a4994b41201]  ETA:[2018-01-02 16:38:16.294506-05:00] 
[2018-01-02 16:38:17,616: INFO/ForkPoolWorker-1] Task celery_workers.workers.simulator.simulate_time[ae12d705-9c7d-4cb5-9192-2a4994b41201] succeeded in 0.0040496350011380855s: 34133
[2018-01-02 16:38:17,622: INFO/MainProcess] Received task: celery_workers.workers.simulator.simulate_time[65bbbdcd-1fd4-437b-ac0a-13e46988cad9]  ETA:[2018-01-02 16:38:17.675494-05:00] 
[2018-01-02 16:38:17,678: INFO/ForkPoolWorker-1] Task celery_workers.workers.simulator.simulate_time[65bbbdcd-1fd4-437b-ac0a-13e46988cad9] succeeded in 0.0011460400019132067s: 34134
[2018-01-02 16:38:17,680: INFO/MainProcess] Received task: celery_workers.workers.simulator.simulate_time[86582f17-6352-42b3-9280-033e36c1095a]  ETA:[2018-01-02 16:38:17.740223-05:00] 
[2018-01-02 16:38:19,614: INFO/ForkPoolWorker-1] Task celery_workers.workers.simulator.simulate_time[86582f17-6352-42b3-9280-033e36c1095a] succeeded in 0.004222234998451313s: 34135
[2018-01-02 16:38:19,615: INFO/MainProcess] Received task: celery_workers.workers.simulator.simulate_time[88cbf278-7930-425f-a87b-cea2b2e6456c]  ETA:[2018-01-02 16:38:19.672533-05:00] 
[2018-01-02 16:38:19,680: INFO/ForkPoolWorker-1] Task celery_workers.workers.simulator.simulate_time[88cbf278-7930-425f-a87b-cea2b2e6456c] succeeded in 0.0008046680013649166s: 34136
[2018-01-02 16:38:19,681: INFO/MainProcess] Received task: celery_workers.workers.simulator.simulate_time[e5338d3b-d0ae-40a7-badf-0b4d8a26540c]  ETA:[2018-01-02 16:38:19.742112-05:00] 
[2018-01-02 16:38:19,779: INFO/ForkPoolWorker-1] Task celery_workers.workers.simulator.simulate_time[e5338d3b-d0ae-40a7-badf-0b4d8a26540c] succeeded in 0.0027749490000132937s: 34137

**Comment by auvipy** [comment]
thanks for feedback. if the issue still exist after 4.2 release, work should be carried out for v5

**Comment by auvipy** [comment]
with celery 4.2?

**Comment by legraina** [comment]
No, still doesn't work as expected. I'm still seeing some delay of half a second between the ETA and the time when the task starts.

**Comment by auvipy** [comment]
ops!

**Comment by auvipy** [comment]
I am considering this as a feature or improvement request then of a bug. just to to what is the current status of the issue on your side after 3 years?

**Comment by legraina** [comment]
The project has ended, so I'm not using it anymore.

**Comment by Finteria** [comment]
@auvipy Is there any plans to fix the issue? The setting worker_timer_precision doesn't work in celery 5.2.7

**Comment by GuyGrinwald** [comment]
@auvipy I would like to upvote this... I had to skip Celery on a personal project since I could not control the accuracy of the execution times using the `eta` feature...

**Comment by auvipy** [comment]
can you please check this article and get back again? https://engineering.instawork.com/celery-eta-tasks-demystified-424b836e4e94
